sycl/test/check_device_code/matrix/matrix-nvptx-tf32-test.cpp

```diff
@@ -0,0 +1,141 @@
+          joint_matrix_load(sg, sub_c, accC.get_pointer(), N);
+
+          // CHECK: tail call i32 @llvm.nvvm.f2tf32.rna(float {{.*}}
+          // Round a, b to tf32
+          for (auto i = 0; i < 4; ++i)
+            sub_a.data[i] = float_to_tf32(sub_a.data[i]);
+
+          for (auto i = 0; i < 4; ++i)
+            sub_b.data[i] = float_to_tf32(sub_b.data[i]);
+
+          //CHECK: tail call { float, float, float, float, float, float, float, float } @llvm.nvvm.wmma.m16n16k8.mma.row.row.tf32(i32 {{.*}}, i32 {{.*}}, i32 {{.*}}, i32 {{.*}}, i32 {{.*}}, i32 {{.*}}, i32 %{{.*}}, i32 {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}) #{{.*}}
+          sub_c = joint_matrix_mad(sg, sub_a, sub_b, sub_c);
+          //CHECK: tail call void @llvm.nvvm.wmma.m16n16k16.store.d.row.stride.f32.p1f32(float addrspace(1)* {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, i32 {{.*}}
+          joint_matrix_store(sg, sub_c, accD.get_pointer(), N);
+        });
+  });
+
+  q.submit([&](handler &cgh) {
+    auto accA = bufA.get_access<access::mode::read_write>(cgh);
+    auto accB = bufB.get_access<access::mode::read_write>(cgh);
+    auto accC = bufC.get_access<access::mode::read_write>(cgh);
+    auto accD = bufD.get_access<access::mode::read_write>(cgh);
+
+    cgh.parallel_for<class col_col>(
+        nd_range<2>({1, 32}, {1, 32}),
+        [=](nd_item<2> item) [[sycl::reqd_work_group_size(1, 1, 32)]] {
+          sycl::sub_group sg = item.get_sub_group();
+
+          joint_matrix<precision::tf32, matrix_use::a, M, K,
+                       matrix_layout::col_major>
+              sub_a;
+
+          joint_matrix<precision::tf32, matrix_use::b, K, N,
+                       matrix_layout::col_major>
+              sub_b;
+
+          joint_matrix<float, matrix_use::accumulator, M, N,
+                       matrix_layout::col_major>
+              sub_c;
+
+          //CHECK: tail call { i32, i32, i32, i32 } @llvm.nvvm.wmma.m16n16k8.load.a.col.stride.tf32.p0i32(i32* %call.ascast.i.i{{.*}}.i, i32 8) #{{.*}}
+          joint_matrix_load(sg, sub_a, accA.get_pointer(), K);
+          //CHECK: tail call { i32, i32, i32, i32 } @llvm.nvvm.wmma.m16n16k8.load.b.col.stride.tf32.p0i32(i32* %call.ascast.i.i{{.*}}.i, i32 16) #{{.*}}
+          joint_matrix_load(sg, sub_b, accB.get_pointer(), N);
+          //CHECK: tail call { float, float, float, float, float, float, float, float } @llvm.nvvm.wmma.m16n16k16.load.c.col.stride.f32.p1f32(float addrspace(1)* {{.*}}, i32 {{.*}}) #{{.*}}
+          joint_matrix_load(sg, sub_c, accC.get_pointer(), N);
+
+          // CHECK: tail call i32 @llvm.nvvm.f2tf32.rna(float {{.*}}
+          // Round a, b to tf32
+          for (auto i = 0; i < 4; ++i)
+            sub_a.data[i] = float_to_tf32(sub_a.data[i]);
```

Comment:
does data contain WI portion of the matrix?
how do you calculate "4" which id the size of the loop?

Reply:
The only dimensions allowed for wmma for tf32 are m=16, n=16, k=8. This means that A, B are 16x8, 8x16 matrices, respectively. wmma operations are executed by a single warp, ie 32 threads. So when A, B are loaded into registers, each thread holds a fragment of each matrix, with dimensions 8x16/32 = 4 elements. 

See here https://github.com/intel/llvm/pull/5870/files/618c80750930b0eaec8cde468c880d52ba54c80c#diff-f71a436bdeda598b29caad471fa637a2844a12f38fe4e85b15b2ccb37bd09833R37 that `data` is the fragment of a matrix which is local to each thread. Notice as well that since the accumulator matrix C is 16x16, the fragments of C have size 8.

Reply:
This sounds good, so this is really implementing the element indexing we added in https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp#L68
Is there a reason why not match the current syntax we have?
Remember, the whole goal is to use the same extension across AMX, DPAS, tensorcores and others coming.

Reply:
> This sounds good, so this is really implementing the element indexing we added in https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp#L68 Is there a reason why not match the current syntax we have? Remember, the whole goal is to use the same extension across AMX, DPAS, tensorcores and others coming.

There's no reason to not match the current syntax you use: we waited to implement the `get_wi_data` stuff until we had a better idea of the element wise operation cases we needed to support (the math functions for example).  Now we understand these cases we are ready for the `get_wi_data` stuff to be implemented which we will do next.

Reply:
The only caveat is that ideally we would add to the wi_data API in the way described here: https://github.com/intel/llvm/pull/5964#issuecomment-1096846407